src/nl/hannahsten/texifyidea/inspections/latex/probablebugs/LatexQuoteInspection.kt

```diff
@@ -0,0 +1,118 @@
+package nl.hannahsten.texifyidea.inspections.latex.probablebugs
+
+import com.intellij.codeInspection.InspectionManager
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiTreeUtil.findChildrenOfType
+import nl.hannahsten.texifyidea.inspections.InsightGroup
+import nl.hannahsten.texifyidea.inspections.TexifyInspectionBase
+import nl.hannahsten.texifyidea.inspections.TexifyRegexInspection
+import nl.hannahsten.texifyidea.psi.LatexNormalText
+import nl.hannahsten.texifyidea.util.files.commandsInFile
+import nl.hannahsten.texifyidea.util.magic.GeneralMagic
+import nl.hannahsten.texifyidea.util.magic.PatternMagic
+import nl.hannahsten.texifyidea.util.magic.RegexPattern
+import nl.hannahsten.texifyidea.util.requiredParameter
+import nl.hannahsten.texifyidea.util.toTextRange
+import java.util.regex.Pattern
+import javax.annotation.RegEx
+
+/**
+ * @author Michael Milton
+ */
+class LatexQuoteInspection : TexifyInspectionBase() {
+    override val inspectionGroup = InsightGroup.LATEX;
+    override val inspectionId = "AsciiQuotes";
+
+    /**
+     * Matches anything that looks like a LaTeX quote
+     */
+    private val quotePattern = """["'`]+""".toRegex();
+
+    /**
+     * Defines how to extract a quote pair from a command
+     */
+    data class QuoteMaker(val command: String, val openParam: Int, val closeParam: Int)
+
+    /**
+     * All the ways that csquotes defines a new quotation mark
+     */
+    private val quoteMakers = arrayOf(
+        QuoteMaker("\\MakeOuterQuote", 0, 0),
+        QuoteMaker("\\MakeInnerQuote", 0, 0),
+        QuoteMaker("\\MakeAutoQuote", 0, 1),
+        QuoteMaker("\\MakeForeignQuote", 1, 2),
+        QuoteMaker("\\MakeBlockQuote", 0, 2),
+        QuoteMaker("\\MakeForeignBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHyphenBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHybridBlockQuote", 1, 3),
+    );
+
+    override fun inspectFile(file: PsiFile, manager: InspectionManager, isOntheFly: Boolean): List<ProblemDescriptor> {
+        // Build up a list of acceptable quote pairs. The ith index in both arrays must contain a matching quote
+        // pair. Additional quote pairs are extracted from the csquotes quote making commands
+        val openers = LinkedHashSet(arrayListOf("`", "``"));
+        val closers = LinkedHashSet(arrayListOf("'", "''"));
+        for (quoteMaker in quoteMakers) {
+            for (command in file.commandsInFile(quoteMaker.command)) {
```

Comment:
Ah cool. I didn't know about that. Speaking of caching, how can I about recomputing everything for each call of this method? Can this be recalculated per paragraph or something instead of the entire text each time?

Reply:
No, because you don't know what changed relative to the last time the inspection was run